src/math/mod.rs

```diff
@@ -29,6 +29,124 @@ pub fn canon_egcd(a: i64, b: i64, c: i64) -> Option<(i64, i64, i64)> {
     }
 }
 
+fn pos_mod(n: i64, m: i64) -> i64 {
+    if n < 0 {
+        n + m
+    } else {
+        n
+    }
+}
+
+fn mod_mul(a: i64, b: i64, m: i64) -> i64 {
+    pos_mod((a as i128 * b as i128 % m as i128) as i64, m)
+}
+
+/// Assuming m >= 2 and exp >= 0, finds base ^ exp % m in logarithmic time
+pub fn mod_exp(mut base: i64, mut exp: i64, m: i64) -> i64 {
+    assert!(m >= 2);
+    assert!(exp >= 0);
+    let mut ans = 1 % m;
+    base = base % m;
+    while exp > 0 {
+        if exp % 2 == 1 {
+            ans = mod_mul(ans, base, m);
+        }
+        base = mod_mul(base, base, m);
+        exp /= 2;
+    }
+    pos_mod(ans, m)
+}
+
+/// Assuming m >= 2, finds multiplicative inverse of n under modulus m
+pub fn mod_inv(n: i64, m: i64) -> i64 {
+    mod_exp(n, m - 2, m)
+}
+
+fn miller_test(n: i64, d: i64, r: i64, a: i64) -> bool {
+    let mut x = mod_exp(a, d, n);
+    if x == 1 || x == n - 1 {
+        return true;
+    }
+    for _ in 0..r {
+        x = mod_mul(x, x, n);
+        if x == n - 1 {
+            return true;
+        }
+    }
+    false
+}
+
+const BASES: [i64; 12] = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37];
+/// Assuming x >= 0, returns true if x is prime
+pub fn is_prime(n: i64) -> bool {
+    assert!(n >= 0);
+    if n <= 1 {
+        return false;
+    }
+    if n <= 3 {
+        return true;
+    }
+    let mut d = n - 1;
+    let mut r = 0;
+    while d % 2 == 0 {
+        d /= 2;
+        r += 1;
+    }
+    for base in BASES.iter() {
+        if *base <= n - 2 && !miller_test(n, d, r, *base) {
+            return false;
+        }
+    }
+    true
+}
+
+fn pollard_rho(n: i64) -> i64 {
+    for a in 1..n {
+        let f = |x| pos_mod(mod_mul(x, x, n) + a, n);
+        for b in 0..n {
+            let mut x = b;
+            let mut y = b;
+            loop {
+                x = f(x);
+                y = f(f(y));
+                let p = num::fast_gcd((x - y).abs(), n);
+                if p > 1 && p < n {
+                    return p;
+                }
+                if x == y {
+                    break;
+                }
+            }
+        }
+    }
+    panic!("No divisor found!");
```

Comment:
Are there ways to guarantee a divisor is found quickly in a contest setting? Is randomization necessary?

Reply:
I don't know much about the theory behind this algorithm, I had a very hard time learning it as there aren't many sources to learn from. Also according to Wikipedia https://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm#Complexity "rigorous analysis of the algorithm remains open". In the article they suggest using the polynomial g(x) = x^2 + 1 as the generator and 2 as the starting value, but those can fail and so you then have to choose something else (either a different starting value or a different polynomial).

Reply:
Hm we don't have to figure this out immediately, but let's leave a comment if we can't prove its usability in something like a contest setting.